### PR TITLE
Fix URI::InvalidURIError whenever `ssl` config is true.

### LINKED ISSUE #102: Specifying `ssl => true` leads to exception "URI::InvalidURIError: bad URI(is not URI?)"
When specifying `ssl => true` (with or without `ca_cert` param), the URL string is not interpolated correctly. Here's the error raised:

```
[2018-06-20T14:18:43,541][WARN ][logstash.filters.elasticsearch] Failed to query elasticsearch for previous event {:index=>"cert", :query=>"@timestamp:*", :event=>#<LogStash::Event:0x7ad689eb>, :error=>#<URI::InvalidURIError: bad URI(is not URI?): https://{:host=>"https://example.us-central1.gcp.cloud.es.io:9243/", :scheme=>"https"}:https>}
```

Notice the URL string appears to contain the full params hash (`hash.to_s`): `https://{:host=>"https://example.us-central1.gcp.cloud.es.io:9243/", :scheme=>"https"}` instead of templating the fields in a string template (e.g. `"%{scheme}://%{host}" % { scheme: 'https', host: '...' }`)

This particular trace was generated by the following configuration (subdomain of the cloud instance changed):

```
bin/logstash --log.level debug -e "input { generator { count =>  3 } }
filter { elasticsearch {
  user => elastic password => '$ES_PWD' hosts => ['https://example.us-central1.gcp.cloud.es.io:9243/']
  index => 'cert' query => '@timestamp:*' sort => '_id' fields => { 'sequence' => 'last_sequence' }
  ssl => true
} }
output { stdout {} elasticsearch {
  user => elastic password => '$ES_PWD' hosts => ['https://example.us-central1.gcp.cloud.es.io:9243/']
  index => 'cert'
} }"
```

## Known workaround

Specify all hosts with their `https://` prefix and do not specify the `ssl` attribute.

## Full stack trace

`#<URI::InvalidURIError: bad URI(is not URI?): https://{:host=>"https://example.us-central1.gcp.cloud.es.io:9243/", :scheme=>"https"}:https>`

```
 "uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/uri/rfc3986_parser.rb:67:in `split'",
 "uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/uri/rfc3986_parser.rb:73:in `parse'",
 "uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/uri/common.rb:227:in `parse'",
 "uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/uri/common.rb:714:in `URI'",
 "org/jruby/RubyMethod.java:127:in `call'",
 "/Users/mat/work/elastic/logstash/vendor/bundle/jruby/2.3.0/gems/faraday-0.9.2/lib/faraday/utils.rb:258:in `URI'",
 "/Users/mat/work/elastic/logstash/vendor/bundle/jruby/2.3.0/gems/faraday-0.9.2/lib/faraday/connection.rb:309:in `url_prefix='",
 "/Users/mat/work/elastic/logstash/vendor/bundle/jruby/2.3.0/gems/faraday-0.9.2/lib/faraday/connection.rb:77:in `initialize'",
 "/Users/mat/work/elastic/logstash/vendor/bundle/jruby/2.3.0/gems/elasticsearch-transport-5.0.4/lib/elasticsearch/transport/transport/http/faraday.rb:38:in `__build_connection'",
 "/Users/mat/work/elastic/logstash/vendor/bundle/jruby/2.3.0/gems/elasticsearch-transport-5.0.4/lib/elasticsearch/transport/transport/base.rb:138:in `block in __build_connections'",
 "org/jruby/RubyArray.java:2486:in `map'",
 "/Users/mat/work/elastic/logstash/vendor/bundle/jruby/2.3.0/gems/elasticsearch-transport-5.0.4/lib/elasticsearch/transport/transport/base.rb:130:in `__build_connections'",
 "/Users/mat/work/elastic/logstash/vendor/bundle/jruby/2.3.0/gems/elasticsearch-transport-5.0.4/lib/elasticsearch/transport/transport/base.rb:40:in `initialize'",
 "/Users/mat/work/elastic/logstash/vendor/bundle/jruby/2.3.0/gems/elasticsearch-transport-5.0.4/lib/elasticsearch/transport/client.rb:114:in `initialize'",
 "/Users/mat/work/elastic/logstash/vendor/bundle/jruby/2.3.0/gems/elasticsearch-transport-5.0.4/lib/elasticsearch/transport.rb:26:in `new'",
 "/Users/mat/work/elastic/plugins/logstash-filter-elasticsearch/lib/logstash/filters/elasticsearch/client.rb:28:in `initialize'",
 "/Users/mat/work/elastic/plugins/logstash-filter-elasticsearch/lib/logstash/filters/elasticsearch.rb:145:in `new_client'",
 "/Users/mat/work/elastic/plugins/logstash-filter-elasticsearch/lib/logstash/filters/elasticsearch.rb:149:in `block in get_client'",
 "/Users/mat/work/elastic/plugins/logstash-filter-elasticsearch/lib/logstash/filters/elasticsearch.rb:149:in `get_client'",
 "/Users/mat/work/elastic/plugins/logstash-filter-elasticsearch/lib/logstash/filters/elasticsearch.rb:93:in `filter'",
 "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/filters/base.rb:145:in `do_filter'",
 "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/filters/base.rb:164:in `block in multi_filter'",
 "org/jruby/RubyArray.java:1734:in `each'",
 "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/filters/base.rb:161:in `multi_filter'",
 "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/filter_delegator.rb:44:in `multi_filter'",
 "(eval):46:in `block in filter_func'",
 "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/pipeline.rb:443:in `filter_batch'",
 "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/pipeline.rb:422:in `worker_loop'",
 "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/pipeline.rb:384:in `block in start_workers'"
```

### PR DESCRIPTION
**Pull request by webmat**
See #102 for more context on this PR.

Still missing:

- [x] Version bump & changelog.
- [x] Documentation for the new behaviour.

I'd like to get a review on this new behaviour first, before I write out the
corresponding documentation.

Closes #102

### CODE REVIEW

**Comment by webmat** [comment]
Ok I added this `validate_client_config!` at the last minute and pushed this, but this was kind of lazy of me. Sorry.

Instead of changing the expectations from 0 to 1 (smelly), I'll move this check to a stateless method on the Client class.

**Comment by webmat** [comment]
Here's the behaviour when someone triggers this conflicting config situation:

```
[2018-06-28T14:07:00,960][ERROR][logstash.filters.elasticsearch] Conflicting configuration detected: you cannot specify a schema in your hosts and set the 'ssl' option at the same time. You must use one or the other. {:ssl=>true, :hosts=>["https://example.us-central1.gcp.cloud.es.io:9243/"]}
[2018-06-28T14:07:00,985][ERROR][logstash.pipeline        ] Error registering plugin {:pipeline_id=>"main", :plugin=>"#<LogStash::FilterDelegator:0x1520c155 @metric_events_out=org.jruby.proxy.org.logstash.instrument.metrics.counter.LongCounter$Proxy2 -  name: out value:0, @metric_events_in=org.jruby.proxy.org.logstash.instrument.metrics.counter.LongCounter$Proxy2 -  name: in value:0, @metric_events_time=org.jruby.proxy.org.logstash.instrument.metrics.counter.LongCounter$Proxy2 -  name: duration_in_millis value:0, @id=\"a0f2c1c31327708709cba79086afa8959198494108fe2c4db284de40ddc05c81\", @klass=LogStash::Filters::Elasticsearch, @metric_events=#<LogStash::Instrument::NamespacedMetric:0x72b8264f @metric=#<LogStash::Instrument::Metric:0x60598838 @collector=#<LogStash::Instrument::Collector:0x34a71b53 @agent=nil, @metric_store=#<LogStash::Instrument::MetricStore:0x279e4efb @store=#<Concurrent::Map:0x00000000000fb0 entries=2 default_proc=nil>, @structured_lookup_mutex=#<Mutex:0x3462253f>, @fast_lookup=#<Concurrent::Map:0x00000000000fb4 entries=62 default_proc=nil>>>>, @namespace_name=[:stats, :pipelines, :main, :plugins, :filters, :a0f2c1c31327708709cba79086afa8959198494108fe2c4db284de40ddc05c81, :events]>, @filter=<LogStash::Filters::Elasticsearch user=>\"elastic\", password=><password>, hosts=>[\"https://ecb5659675444fdea0ec2129b65410c0.us-central1.gcp.cloud.es.io:9243/\"], index=>\"cert\", query=>\"@timestamp:*\", sort=>\"_id\", fields=>{\"sequence\"=>\"last_sequence\"}, ssl=>true, id=>\"a0f2c1c31327708709cba79086afa8959198494108fe2c4db284de40ddc05c81\", enable_metric=>true, periodic_flush=>false, enable_sort=>true, result_size=>1, tag_on_failure=>[\"_elasticsearch_lookup_failure\"]>>", :error=>"Aborting due to conflicting configuration", :thread=>"#<Thread:0x4921c1ba run>"}
[2018-06-28T14:07:01,002][ERROR][logstash.pipeline        ] Pipeline aborted due to error {:pipeline_id=>"main", :exception=>#<LogStash::ConfigurationError: Aborting due to conflicting configuration>, :backtrace=>["/Users/mat/work/elastic/plugins/logstash-filter-elasticsearch/lib/logstash/filters/elasticsearch/client.rb:24:in `initialize'", "/Users/mat/work/elastic/plugins/logstash-filter-elasticsearch/lib/logstash/filters/elasticsearch.rb:143:in `new_client'", "/Users/mat/work/elastic/plugins/logstash-filter-elasticsearch/lib/logstash/filters/elasticsearch.rb:147:in `block in get_client'", "/Users/mat/work/elastic/plugins/logstash-filter-elasticsearch/lib/logstash/filters/elasticsearch.rb:147:in `get_client'", "/Users/mat/work/elastic/plugins/logstash-filter-elasticsearch/lib/logstash/filters/elasticsearch.rb:74:in `register'", "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/pipeline.rb:340:in `register_plugin'", "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/pipeline.rb:351:in `block in register_plugins'", "org/jruby/RubyArray.java:1734:in `each'", "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/pipeline.rb:351:in `register_plugins'", "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/pipeline.rb:729:in `maybe_setup_out_plugins'", "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/pipeline.rb:361:in `start_workers'", "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/pipeline.rb:288:in `run'", "/Users/mat/work/elastic/logstash/logstash-core/lib/logstash/pipeline.rb:248:in `block in start'"], :thread=>"#<Thread:0x4921c1ba run>"}
[2018-06-28T14:07:01,034][ERROR][logstash.agent           ] Failed to execute action {:id=>:main, :action_type=>LogStash::ConvergeResult::FailedAction, :message=>"Could not execute action: PipelineAction::Create<main>, action_result: false", :backtrace=>nil}
...
[2018-06-28T14:07:06,309][DEBUG][logstash.agent           ] Shutting down all pipelines {:pipelines_count=>1}
```

Actually should I be using `FATAL`, since I'm basically crashing Logstash here?

File: lib/logstash/filters/elasticsearch/client.rb [hunk 1]
```diff
@@ -32,6 +36,20 @@ def search(params)
         @client.search(params)
       end
 
+      def self.validate_config!(user, password, options={})
+        ssl     = options[:ssl]
+        hosts   = options[:hosts]
+        logger  = options[:logger]
+        unless ssl.nil?
+          if hosts.detect {|h| h =~ %r(^https?://) }
```
Comment (yaauie) [inline lib/logstash/filters/elasticsearch/client.rb]: a couple things here:
 - mixing `unless` and `if` in nested blocks is a bit confusing to me
 - while `ssl` will be nil at this point if `options` didn't include the `:ssl` directive, that is more of an _affect_ and may be clearer to test `options` for `:ssl` directly.
 - `String#start_with?(*other)` is more readable than regex (and faster to execute, too).

consider:

~~~ ruby
        if options.include?(:ssl) && hosts.any? { |h| h.start_with?('http://', 'https://') }
          logger.error # ...
~~~

**Comment by webmat** [comment]
@yaauie Thanks for the feedback. I was initially trying to avoid adding too much subtle logic to this, since we want to rewrite/harmonize the HTTP clients between the 3 ES plugins. Keep in mind that `ssl => true` currently doesn't work at all (see exception in #102).

But after your feedback I looked again at output-es, and what you're suggesting is in line with the behaviour over there, and much less brutal than my implementation 😂 So I'll change the validation as you suggest.

My checklist:

- [ ] Simplify if/unless, use `start_with?`
- [ ] Harmonize schemes as long as they don't conflict

For reference, here's the implementation of SSL/schema in output-es:

* [config part 1](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/common_configs.rb#L93]
* [config part 2](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch.rb#L136-L154)
* [ssl handling part 1](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/http_client.rb#L212-L283)
* [ssl handling part 2](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/http_client_builder.rb#L106-L141)

**Comment by kostasb** [comment]
@webmat Are there any plans to merge this PR?
Currently SSL configuration works the opposite way than [documented](https://www.elastic.co/guide/en/logstash/6.6/plugins-filters-elasticsearch.html) with the https protocol having to be specified in the hosts list and if I understand correctly this PR aligns the configuration to be as expected per the documentation.

**Comment by webmat** [comment]
@kostasb Sorry this is from around the time I left the Logstash team. I tried getting it reviewed a few times after that, and then forgot about it.

Hey @jsvd or @yaauie, would you mind picking this up?

**Comment by R0ky** [comment]
Hey @jsvd or @yaauie, just wondering if there is any news regarding this PR?

**Comment by renshuki** [comment]
@jsvd, @yaauie is there any blocker to merge this? This `ssl => true` bug has been there for a while and need to be fixed. It currently leads to a lot of confusion for the users. Thanks.

**Comment by edmocosta** [comment]
Closing this pull request based on the comment: https://github.com/logstash-plugins/logstash-filter-elasticsearch/issues/102#issuecomment-1697444353. Feel free to reopen if the problem still exists. Thank you.

**Pull request closed without merging**